### PR TITLE
[WIP] provider/aws: Add aws_workspace resource

### PR DESCRIPTION
**Pull request by liamjbennett**
- [x] CRUD
- [x] Exported Attributes
- [x] Timeouts (> 20 mins)
- [x] Acceptance Tests
- [x] Docs

### CODE REVIEW

**Comment by liamjbennett** [comment]
Question: how do I make the resource work with the `count` parameter?

**Comment by radeksimko** [comment]
@liamjbennett It should just work out of the box - that's a terraform's core feature, not resource specific thing.

The only thing to keep in mind is that you can replicate _resources_, but _not_ inner nested blocks inside resources. e.g. you can use `count` for `aws_elb` to create as many ELBs as you want, but you cannot use `count` for `aws_elb.listeners` to create N listeners for a given ELB.

**Comment by liamjbennett** [comment]
TODO: 
  - `bundle_id` - make this optional and allow lookup via `bundle_name` for bundles provided via Amazon
  - investigate creating the `aws_workspace_bundle` resource to support custom bundles.

File: builtin/providers/aws/resource_aws_workspace.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	if v, ok := d.GetOk("bundle_id"); ok {
+		params.BundleId = aws.String(v.(string))
+	} else {
+		// Only looking up from the default provided AMAZON bundles at this point
+		foundBundles, err := findWorkspacesBundle("AMAZON", d.Get("bundle_name").(string), meta)
+
+		if err != nil {
+			log.Printf("Error finding bundle: %q", err.Error())
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_workspace.go]: we can wrap this log and return err into

```
return fmt.Errorf("Error finding bundle: %q"m err)
```

File: builtin/providers/aws/resource_aws_workspace.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+			return err
+		}
+
+		params.BundleId = foundBundles[0].BundleId
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_workspace.go]: Are you 100% sure it's the first bundle in the collection you want to use?

File: builtin/providers/aws/resource_aws_workspace.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+		Delete: resourceAwsWorkspaceDelete,
+
+		Schema: map[string]*schema.Schema{
+			"bundle_name": &schema.Schema{
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+
+			"bundle_id": &schema.Schema{
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_workspace.go]: If you don't set this and then it get's set in the Read func, you will get a perpetual diff. As this is 'ForceNew: true', that could be problematic. I think this will also need to be `Computed: true`

File: builtin/providers/aws/tagsWorkspace.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/workspaces"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+// setTags is a helper to set the tags for a resource. It expects the
+// tags field to be named "tags"
+func setTagsWorkspace(conn *workspaces.WorkSpaces, d *schema.ResourceData) error {
+	return nil
```
Comment (stack72) [inline builtin/providers/aws/tagsWorkspace.go]: We don't actually set any tags? If not, why is this here?

File: builtin/providers/aws/resource_aws_workspace.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+	})
+
+	if err != nil {
+		log.Printf("[ERROR] Error finding Workspace: %s", err)
+		return err
+	}
+
+	// If nothing was found, then return no state
+	if len(resp.Workspaces) == 0 {
+		d.SetId("")
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_workspace.go]: We should drop in a log.Printf message to say that nothing was found and we are removing from state. This will help people debug why things disappear :)

File: builtin/providers/aws/resource_aws_workspace.go [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+func resourceAwsWorkspaceRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	resp, err := conn.DescribeWorkspaces(&workspaces.DescribeWorkspacesInput{
+		WorkspaceIds: []*string{aws.String(d.Id())},
+	})
+
+	if err != nil {
+		log.Printf("[ERROR] Error finding Workspace: %s", err)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_workspace.go]: combine to fmt.ErrorF

**Comment by stack72** [comment]
Hey @liamjbennett 

A few small comments inline. Nothing major.

Please can you drop the results of the test into the PR as well for reference :)

P.

**Comment by liamjbennett** [comment]
@stack72 @radeksimko
Getting an error when running the first acceptence test:

```panic: interface conversion: interface {} is nil, not *aws.AWSClient``` 

[Full Gist](https://gist.github.com/liamjbennett/801fe276d0ee51af516f388071c4bcf9)

**Comment by radeksimko** [comment]
@liamjbennett Have a look at the diff of `builtin/providers/aws/provider.go`:

```diff
 -		ConfigureFunc: providerConfigure,
```

That's most likely the cause of that error 😉

**Comment by liamjbennett** [comment]
@radeksimko  Doh. Yes that solved the issue.

I am now getting a genuine test failure - which I beleieve might be something to do with the active directory dependency. Is there anyway to see the TF run that is executed as part of the test?

**Comment by radeksimko** [comment]
> Is there anyway to see the TF run that is executed as part of the test?

You mean to see the debug output from tests? Setting the right ENV variables may help: https://github.com/hashicorp/terraform/blob/master/website/source/docs/internals/debugging.html.md#debugging-terraform

**Comment by liamjbennett** [comment]
This is currently blocked because the workspace api support registering the directory e.g.

https://docs.aws.amazon.com/workspaces/latest/adminguide/registration.html

**Comment by liamjbennett** [comment]
Ref: https://github.com/aws/aws-sdk-go/issues/814

**Comment by liamjbennett** [comment]
Given that we are unable to test this due to the missing api's and the total silence from aws on the matter I am going to close this PR for now and come back to it at a future point in time.

**Comment by jayyvis** [comment]
@liamjbennett 
If directory registration API is the only missing piece, then how about separating it out of workspace provisioning?

Cloud Formation currently handles it that way.

> Before creating a Workspace in CloudFormation, you must register a Directory Service directory with Workspaces.

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-workspaces-workspace.html

**Comment by liamjbennett** [comment]
@jayyvis yes we do that with cloud formation also but for the sake of this PR there is no way to actually run the tests without manually performing that registration, which is not acceptable.

**Comment by wmantly** [comment]
A year on, is there any way to get this done now?

**Comment by apparentlymart** [comment]
Since the AWS provider is no longer developed in this repository, the tracking issue for this is terraform-providers/terraform-provider-aws#434 . There will be no further updates posted here because the relevant code is no longer in this repository.
